cranelift-codegen/meta/src/cdsl/instructions.rs

```diff
@@ -408,36 +410,126 @@ impl ValueTypeOrAny {
+    ImmediateValue(Immediate),
+}
+
+impl From<shared::types::Int> for InstructionParameter {
+    fn from(ty: shared::types::Int) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Bool> for InstructionParameter {
+    fn from(ty: shared::types::Bool) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Float> for InstructionParameter {
+    fn from(ty: shared::types::Float) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<cdsl::types::LaneType> for InstructionParameter {
+    fn from(ty: cdsl::types::LaneType) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Reference> for InstructionParameter {
+    fn from(ty: shared::types::Reference) -> Self {
+        InstructionParameter::ReferenceType(ty.into())
+    }
+}
+
+#[derive(Clone)]
+pub enum Immediate {
+    IntCC(u8),
+    FloatCC(u8),
+    UInt8(u8),
+    UInt128(u128),
+}
+
 #[derive(Clone)]
 pub struct BoundInstruction {
     pub inst: Instruction,
     pub value_types: Vec<ValueTypeOrAny>,
+    pub immediate_values: Vec<Immediate>,
 }
 
 impl BoundInstruction {
-    pub fn bind(self, lane_type: impl Into<LaneType>) -> BoundInstruction {
-        bind(self.inst, Some(lane_type.into()), self.value_types)
+    /// Construct a new bound instruction (with nothing bound yet) from an instruction
+    fn new(inst: &Instruction) -> Self {
+        BoundInstruction {
+            inst: inst.clone(),
+            value_types: vec![],
+            immediate_values: vec![],
+        }
     }
 
-    pub fn bind_ref(self, reference_type: impl Into<ReferenceType>) -> BoundInstruction {
-        bind_ref(self.inst, Some(reference_type.into()), self.value_types)
+    /// Verify that binding types to the instruction does not violate the polymorphic rules.
+    fn verify_polymorphic_binding(&self) -> Result<(), String> {
+        match &self.inst.polymorphic_info {
+            Some(poly) => {
+                if self.value_types.len() <= 1 + poly.other_typevars.len() {
+                    Ok(())
+                } else {
+                    Err(format!(
+                        "trying to bind too many types for {}",
+                        self.inst.name
+                    ))
+                }
+            }
+            None => Err(format!(
+                "trying to bind a type for {} which is not a polymorphic instruction",
+                self.inst.name
+            )),
+        }
     }
 
-    pub fn bind_vector_from_lane(
-        self,
-        lane_type: impl Into<LaneType>,
-        vector_size_in_bits: u64,
-    ) -> BoundInstruction {
-        bind_vector(
-            self.inst,
-            lane_type.into(),
-            vector_size_in_bits,
-            self.value_types,
-        )
+    fn verify_immediate_binding(&self) -> Result<(), String> {
+        // TODO
+        Ok(())
```

Comment:
Not exactly sure how I would need to check this but I believe I need to ensure at the very least that I don't have more immediates to bind than the format specifies (but again I might need a `FormatRegistry` for that because the `Instruction` only has a format index).